File: examples/clj/src/com/rpl/agent/basic/tools_agent.clj
(ns com.rpl.agent.basic.tools-agent
  "Demonstrates LangChain4j tools integration with OpenAI chat models.

  Features demonstrated:
  - new-tools-agent: Create specialized agent for tool execution
  - tool-specification: Define tool schemas for LangChain4j
  - OpenAI model with tool calling capabilities
  - Natural language to tool execution workflow"
  (:require
   [clojure.string :as str]
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.agent-o-rama.langchain4j :as lc4j]
   [com.rpl.agent-o-rama.langchain4j.json :as lj]
   [com.rpl.agent-o-rama.tools :as tools]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest])
  (:import
   [dev.langchain4j.data.message
    ToolExecutionResultMessage
    UserMessage]
   [dev.langchain4j.model.openai
    OpenAiChatModel]))

;;; Tool function definitions
(defn calculate-tool
  "Simple calculator tool that performs basic arithmetic"
  [args]
  (let [operation (args "operation")
        a         (args "a")
        b         (args "b")
        result    (case operation
                    "add" (+ a b)
                    "subtract" (- a b)
                    "multiply" (* a b)
                    "divide" (if (zero? b)
                               "Error: Division by zero"
                               (/ a b))
                    "Error: Unknown operation")]
    (str result)))

(defn string-tool
  "String manipulation tool for text processing"
  [args]
  (let [text      (args "text")
        operation (args "operation")]
    (case operation
      "uppercase" (str/upper-case text)
      "lowercase" (str/lower-case text)
      "reverse" (str/reverse text)
      "length" (str (count text))
      "Error: Unknown string operation")))

;;; Tool specifications for LangChain4j
(def CALCULATOR-TOOL
  (tools/tool-info
   (tools/tool-specification
    "calculator"
    (lj/object
     {:description "Parameters for calculator operations"
      :required    ["operation" "a" "b"]}
     {"operation" (lj/enum "The arithmetic operation to perform"
                           ["add" "subtract" "multiply" "divide"])
      "a"         (lj/number "The first number")
      "b"         (lj/number "The second number")})
    "Performs basic arithmetic operations on two numbers")
   calculate-tool))

(def STRING-TOOL
  (tools/tool-info
   (tools/tool-specification
    "string-processor"
    (lj/object
     {:description "Parameters for string manipulation operations"
      :required    ["text" "operation"]}
     {"text"      (lj/string "The text to process")
      "operation" (lj/enum "The string operation to perform"
                           ["uppercase" "lowercase" "reverse" "length"])})
    "Performs string manipulation operations")
   string-tool))

;;; Agent module demonstrating tools functionality
(aor/defagentmodule ToolsAgentModule
  [topology]

  ;; Declare OpenAI model
  (aor/declare-agent-object-builder
   topology
   "openai-model"
   (fn [_setup]
     (-> (OpenAiChatModel/builder)
         (.apiKey (or (System/getenv "OPENAI_API_KEY") "fake-key-for-demo"))
         (.modelName "gpt-4o-mini")
         .build)))

  ;; Create tools agent with our tool definitions
  (tools/new-tools-agent
   topology
   "ToolsAgent"
   [CALCULATOR-TOOL STRING-TOOL])

  ;; Create a coordinator agent that uses OpenAI with tools
  (->
    topology
    (aor/new-agent "ToolsCoordinator")

    ;; Node that sends natural language prompts to OpenAI and processes tool calls
    (aor/node
     "chat-with-tools"
     nil
     (fn chat-with-tools-fn [agent-node prompts]
       (let [model       (aor/get-agent-object agent-node "openai-model")
             tools-agent (aor/agent-client agent-node "ToolsAgent")
             results     (atom [])]

         (doseq [^String prompt prompts]
           ;; Send prompt to OpenAI model with tools
           (let [response   (lc4j/chat model
                                       (lc4j/chat-request
                                        [(UserMessage. prompt)]
                                        {:tools [CALCULATOR-TOOL STRING-TOOL]}))
                 ai-message (.aiMessage response)
                 tool-calls (vec (.toolExecutionRequests ai-message))]

             (if (seq tool-calls)
               ;; Execute tools and get results
               (let [tool-results (aor/agent-invoke tools-agent tool-calls)]
                 (swap! results conj
                   {:prompt       prompt
                    :tool-calls   (count tool-calls)
                    :tool-results tool-results}))
               (swap! results conj
                 {:prompt     prompt
                  :response   (.text ai-message)
                  :tool-calls 0}))))

         (aor/result! agent-node
                      {:prompts-count (count prompts)
                       :results       @results}))))))

(defn -main
  "Run the tools agent example"
  [& _args]
  (with-open [ipc (rtest/create-ipc)]
    (rtest/launch-module! ipc ToolsAgentModule {:tasks 1 :threads 1})
    (let [manager     (aor/agent-manager
                       ipc
                       (rama/get-module-name ToolsAgentModule))
          coordinator (aor/agent-client manager "ToolsCoordinator")
          prompts     ["What is 15 plus 25?"
                       "Calculate 7 times 8"
                       "Divide 100 by 4"
                       "Convert 'Hello World' to uppercase"
                       "Reverse the text 'ReverseMe'"
                       "How many characters are in 'Count Characters'?"]
          result      (aor/agent-invoke coordinator prompts)]

      ;; Create natural language prompts that will trigger tool usage
      (println "Prompts processed:" (:prompts-count result))

      (println "\nDetailed results:")
      (doseq [[idx prompt-result] (map-indexed vector (:results result))]
        (println (format "\n[%d] Prompt: \"%s\"" (inc idx) (:prompt prompt-result)))
        (if (> (:tool-calls prompt-result 0) 0)
          (do
            (println (format "    Tool calls: %d" (:tool-calls prompt-result)))
            (println "    Tool results:"
                     (mapv
                      #(.text ^ToolExecutionResultMessage %)
                      (:tool-results prompt-result))))
          (println "    Direct response:" (:response prompt-result)))))))

(comment
  (-main))

File: examples/java/basic/src/main/java/com/rpl/agent/basic/ToolsAgent.java
package com.rpl.agent.basic;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentManager;
import com.rpl.agentorama.AgentNode;
import com.rpl.agentorama.AgentTopology;
import com.rpl.agentorama.AgentsModule;
import com.rpl.agentorama.ops.RamaVoidFunction2;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import dev.langchain4j.agent.tool.Tool;
import dev.langchain4j.model.chat.ChatModel;
import dev.langchain4j.model.openai.OpenAiChatModel;
import dev.langchain4j.service.AiServices;

/**
 * Java example demonstrating LangChain4j tools integration with OpenAI chat models.
 *
 * <p>Features demonstrated:
 *
 * <ul>
 *   <li>Tool definition with @Tool annotation
 *   <li>OpenAI model with tool calling capabilities
 *   <li>Natural language to tool execution workflow
 *   <li>AI Services integration
 * </ul>
 *
 * <p>This example requires OPENAI_API_KEY environment variable to be set.
 */
public class ToolsAgent {

  /** Calculator tool for basic arithmetic operations. */
  public static class Calculator implements java.io.Serializable {
    private static final long serialVersionUID = 1L;

    @Tool("Adds two numbers together")
    public double add(double a, double b) {
      return a + b;
    }

    @Tool("Subtracts second number from first")
    public double subtract(double a, double b) {
      return a - b;
    }

    @Tool("Multiplies two numbers")
    public double multiply(double a, double b) {
      return a * b;
    }

    @Tool("Divides first number by second")
    public double divide(double a, double b) {
      if (b == 0) {
        throw new IllegalArgumentException("Cannot divide by zero");
      }
      return a / b;
    }
  }

  /** AI Assistant interface for tool calling. */
  public interface MathAssistant {
    String chat(String userMessage);
  }

  /** Agent Module demonstrating tools functionality. */
  public static class ToolsModule extends AgentsModule {

    @Override
    protected void defineAgents(AgentTopology topology) {
      // Declare OpenAI API key
      topology.declareAgentObject("openai-api-key", System.getenv("OPENAI_API_KEY"));

      // Build OpenAI chat model with configuration
      topology.declareAgentObjectBuilder(
          "openai-model",
          setup -> {
            String apiKey = (String) setup.getAgentObject("openai-api-key");
            return OpenAiChatModel.builder().apiKey(apiKey).modelName("gpt-4o-mini").build();
          });

      // Declare calculator tool
      topology.declareAgentObject("calculator", new Calculator());

      topology.newAgent("ToolsAgent").node("chat-with-tools", null, new ChatWithToolsFunction());
    }
  }

  /** Node function that processes prompts using OpenAI with tools. */
  public static class ChatWithToolsFunction implements RamaVoidFunction2<AgentNode, String> {

    @Override
    public void invoke(AgentNode agentNode, String prompt) {
      ChatModel model = (ChatModel) agentNode.getAgentObject("openai-model");
      Calculator calculator = (Calculator) agentNode.getAgentObject("calculator");

      // Create AI service with tools
      MathAssistant assistant =
          AiServices.builder(MathAssistant.class)
              .chatModel(model)
              .tools(calculator)
              .build();

      // Process prompt and return response
      String response = assistant.chat(prompt);
      agentNode.result(response);
    }
  }

  public static void main(String[] args) throws Exception {
    String apiKey = System.getenv("OPENAI_API_KEY");
    if (apiKey == null || apiKey.trim().isEmpty()) {
      System.out.println("Tools Agent Example:");
      System.out.println("OPENAI_API_KEY environment variable not set.");
      System.out.println("Please set your OpenAI API key to run this example:");
      System.out.println("  export OPENAI_API_KEY=your-api-key-here");
      return;
    }

    try (InProcessCluster ipc = InProcessCluster.create()) {
      ToolsModule module = new ToolsModule();
      ipc.launchModule(module, new LaunchConfig(1, 1));

      String moduleName = module.getModuleName();
      AgentManager manager = AgentManager.create(ipc, moduleName);
      AgentClient agent = manager.getAgentClient("ToolsAgent");

      System.out.println("Tools Agent Example:");
      System.out.println("Using LangChain4j tools with OpenAI...\n");

      // Test with different prompts that trigger tool usage
      String[] prompts = {
        "What is 15 plus 25?",
        "Calculate 7 times 8",
        "Divide 100 by 4",
        "What's the result of 50 minus 12?"
      };

      for (String prompt : prompts) {
        System.out.println("User: " + prompt);
        String result = (String) agent.invoke(prompt);
        System.out.println("Assistant: " + result);
        System.out.println();
      }

      System.out.println("Notice how:");
      System.out.println("- OpenAI automatically calls appropriate tools");
      System.out.println("- Natural language is translated to tool execution");
      System.out.println("- Results are formatted in natural language responses");
    }
  }
}
